File: Week02/94. 二叉树的中序遍历.py
# 94. 二叉树的中序遍历.py
from typing import List


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def inorderTraversal(self, root: TreeNode) -> List[int]:
        res = []
        self._helper(root, res)
        return res

    def _helper(self, node: TreeNode, res: List[int]) -> None:
        if not node:
            return

        self._helper(node.left, res)
        res.append(node.val)
        self._helper(node.right, res)


def main():
    pass


if __name__ == '__main__':
    main()

File: Week02/94. 二叉树的中序遍历 .cpp
//
// Created by Ruizhe Hou on 2020/6/17.
//

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
 // 递归的方法
 class Solution{
 public:
     vector<int> inorderTraversal(TreeNode* root) {
         vector<int> nodes;
         inorder()
     }
 };

// 迭代的方法，使用栈
class Solution {
public:
    vector<int> inorderTraversal(TreeNode *root) {
        vector<int> nodes;
        stack < TreeNode * > todo;
        while (root || !todo.empty()) {
            while (root) {
                todo.push(root);
                root = root->left;
            }
            root = todo.top();
            todo.pop();
            nodes.push_back(root->val);
            root = root->right;
        }
        return nodes;
    }
};
